File: lua/oklch.lua
local function srgbLinearize(v)
  if v <= 0.04045 then return v / 12.92
  else return ((v + 0.055) / 1.055) ^ 2.4 end
end

local function srgbDelinearize(v)
  if v <= 0.0031308 then return v * 12.92
  else return 1.055 * v ^ (1.0 / 2.4) - 0.055 end
end

local function linearSrgbToOklab(r, g, b)
  local l = 0.4122214708 * r + 0.5363325363 * g + 0.0514459929 * b
  local m = 0.2119034982 * r + 0.6806995451 * g + 0.1073969566 * b
  local s = 0.0883024619 * r + 0.2817188376 * g + 0.6299787005 * b
  local lc = l ^ (1 / 3)
  local mc = m ^ (1 / 3)
  local sc = s ^ (1 / 3)
  return
    0.2104542553 * lc + 0.7936177850 * mc - 0.0040720468 * sc,
    1.9779984951 * lc - 2.4285922050 * mc + 0.4505937099 * sc,
    0.0259040371 * lc + 0.7827717662 * mc - 0.8086757660 * sc
end

local function oklabToLinearSrgb(L, a, b)
  local lc = L + 0.3963377774 * a + 0.2158037573 * b
  local mc = L - 0.1055613458 * a - 0.0638541728 * b
  local sc = L - 0.0894841775 * a - 1.2914855480 * b
  local l = lc * lc * lc
  local m = mc * mc * mc
  local s = sc * sc * sc
  return
    4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s,
    -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s,
    -0.0041960863 * l - 0.7034186147 * m + 1.7076147010 * s
end

local function inSrgbGamut(r, g, b)
  local eps = 1e-4
  return r >= -eps and r <= 1 + eps and g >= -eps and g <= 1 + eps and b >= -eps and b <= 1 + eps
end

local M = {}

function M.oklchToHex(hue, chroma, L)
  local hRad = hue * math.pi / 180
  local lo, hi = 0.0, chroma
  local r, g, b
  for _ = 1, 24 do
    local mid = (lo + hi) / 2
    r, g, b = oklabToLinearSrgb(L, mid * math.cos(hRad), mid * math.sin(hRad))
    if inSrgbGamut(r, g, b) then lo = mid else hi = mid end
  end
  r, g, b = oklabToLinearSrgb(L, lo * math.cos(hRad), lo * math.sin(hRad))
  return string.format("#%02x%02x%02x",
    math.floor(srgbDelinearize(math.max(0, math.min(1, r))) * 255 + 0.5),
    math.floor(srgbDelinearize(math.max(0, math.min(1, g))) * 255 + 0.5),
    math.floor(srgbDelinearize(math.max(0, math.min(1, b))) * 255 + 0.5)
  )
end

function M.hexToOklch(hex)
  hex = hex:gsub("#", "")
  local r, g, b
  if #hex == 3 then
    r = tonumber(hex:sub(1, 1):rep(2), 16) / 255
    g = tonumber(hex:sub(2, 2):rep(2), 16) / 255
    b = tonumber(hex:sub(3, 3):rep(2), 16) / 255
  elseif #hex == 6 then
    r = tonumber(hex:sub(1, 2), 16) / 255
    g = tonumber(hex:sub(3, 4), 16) / 255
    b = tonumber(hex:sub(5, 6), 16) / 255
  else
    error("unexpected hex " .. hex)
  end
  local L, a, bv = linearSrgbToOklab(srgbLinearize(r), srgbLinearize(g), srgbLinearize(b))
  local chroma = math.sqrt(a * a + bv * bv)
  local hue = math.atan2(bv, a) * 180 / math.pi
  if hue < 0 then hue = hue + 360 end
  return { hue = hue, chroma = chroma, L = L }
end

return M

File: typescript/oklch.ts
// OKLCH color space implementation.
const srgbLinearize = (v: number) => (v <= 0.04045 ? v / 12.92 : ((v + 0.055) / 1.055) ** 2.4)

const srgbDelinearize = (v: number) => (v <= 0.0031308 ? v * 12.92 : 1.055 * v ** (1.0 / 2.4) - 0.055)

const linearSrgbToOklab = (r: number, g: number, b: number) => {
  const l = 0.4122214708 * r + 0.5363325363 * g + 0.0514459929 * b
  const m = 0.2119034982 * r + 0.6806995451 * g + 0.1073969566 * b
  const s = 0.0883024619 * r + 0.2817188376 * g + 0.6299787005 * b
  const lc = l ** (1 / 3),
    mc = m ** (1 / 3),
    sc = s ** (1 / 3)
  return [
    0.2104542553 * lc + 0.793617785 * mc - 0.0040720468 * sc,
    1.9779984951 * lc - 2.428592205 * mc + 0.4505937099 * sc,
    0.0259040371 * lc + 0.7827717662 * mc - 0.808675766 * sc
  ]
}

const oklabToLinearSrgb = (L: number, a: number, b: number) => {
  const lc = L + 0.3963377774 * a + 0.2158037573 * b
  const mc = L - 0.1055613458 * a - 0.0638541728 * b
  const sc = L - 0.0894841775 * a - 1.291485548 * b
  const l = lc ** 3,
    m = mc ** 3,
    s = sc ** 3
  return [
    4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s,
    -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s,
    -0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s
  ]
}

const inSrgbGamut = (r: number, g: number, b: number) => {
  const eps = 1e-4
  return r >= -eps && r <= 1 + eps && g >= -eps && g <= 1 + eps && b >= -eps && b <= 1 + eps
}

const toHex = (v: number) =>
  Math.floor(srgbDelinearize(Math.max(0, Math.min(1, v))) * 255 + 0.5)
    .toString(16)
    .padStart(2, '0')

export const oklchToHex = (hue: number, chroma: number, L: number) => {
  const hRad = (hue * Math.PI) / 180
  let lo = 0,
    hi = chroma
  for (let i = 0; i < 24; i++) {
    const mid = (lo + hi) / 2
    const [r, g, b] = oklabToLinearSrgb(L, mid * Math.cos(hRad), mid * Math.sin(hRad))
    if (inSrgbGamut(r, g, b)) lo = mid
    else hi = mid
  }
  const [r, g, b] = oklabToLinearSrgb(L, lo * Math.cos(hRad), lo * Math.sin(hRad))
  return `#${toHex(r)}${toHex(g)}${toHex(b)}`
}

export const hexToOklch = (hex: string) => {
  hex = hex.replace('#', '')
  let r: number, g: number, b: number
  if (hex.length == 3) {
    r = parseInt(hex[0] + hex[0], 16) / 255
    g = parseInt(hex[1] + hex[1], 16) / 255
    b = parseInt(hex[2] + hex[2], 16) / 255
  } else {
    r = parseInt(hex.slice(0, 2), 16) / 255
    g = parseInt(hex.slice(2, 4), 16) / 255
    b = parseInt(hex.slice(4, 6), 16) / 255
  }
  const [L, a, bv] = linearSrgbToOklab(srgbLinearize(r), srgbLinearize(g), srgbLinearize(b))
  const chroma = Math.sqrt(a * a + bv * bv)
  let hue = (Math.atan2(bv, a) * 180) / Math.PI
  if (hue < 0) hue += 360
  return { hue, chroma, L }
}
